contracts/modules/Experimental/TransferManager/SignedTransferManager.sol

```diff
@@ -0,0 +1,207 @@
+    */
+    function checkSignatureValidity(bytes calldata _data) external view returns(bool) {
+        address targetAddress;
+        uint256 nonce;
+        uint256 expiry;
+        bytes memory signature;
+        (targetAddress, nonce, expiry, signature) = abi.decode(_data, (address, uint256, uint256, bytes));
+        if (targetAddress != address(this) || expiry < now || signature.length == 0 || _checkSignatureIsInvalid(signature))
+            return false;
+        return true;
+    }
+
+    function checkSigner(address _signer) external view returns(bool) {
+        return _checkSigner(_signer);
+    }
+
+    /**
+    * @notice function to remove or add signer(s) onto the signer mapping
+    * @param _signers address array of signers
+    * @param _signersStats bool array of signers stats
+    */
+    function updateSigners(address[] calldata _signers, bool[] calldata _signersStats) external withPerm(ADMIN) {
+        require(_signers.length == _signersStats.length, "Array length mismatch");
+        IDataStore dataStore = IDataStore(ISecurityToken(securityToken).dataStore());
+        for(uint256 i=0; i<_signers.length; i++) {
+            require(_signers[i] != address(0), "Invalid address");
+            dataStore.setBool(keccak256(abi.encodePacked(SIGNER, _signers[i])), _signersStats[i]);
+        }
+        emit SignersUpdated(_signers, _signersStats);
+    }
+
+    /**
+    * @notice allow verify transfer with signature
+    * @param _from address transfer from
+    * @param _to address transfer to
+    * @param _amount transfer amount
+    * @param _data signature
+    * @param _isTransfer bool value of isTransfer
+    * Sig needs to be valid (not used or deemed as invalid)
+    * Signer needs to be in the signers mapping
+    */
+    function verifyTransfer(address _from, address _to, uint256 _amount, bytes memory _data , bool _isTransfer) public returns(Result) {
+        if (!paused) {
+
+            require (_isTransfer == false || msg.sender == securityToken, "Sender is not ST");
+
+            if (_data.length == 0)
+                return Result.NA;
+
+            address targetAddress;
+            uint256 nonce;
+            uint256 expiry;
+            bytes memory signature;
+            (targetAddress, nonce, expiry, signature) = abi.decode(_data, (address, uint256, uint256, bytes));
+
+            if (address(this) != targetAddress || signature.length == 0 || _checkSignatureIsInvalid(signature) || expiry < now)
+                return Result.NA;
+
+            bytes32 hash = keccak256(abi.encodePacked(targetAddress, nonce, expiry, _from, _to, _amount));
+            bytes32 prependedHash = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash));
+            address signer = _recoverSignerAdd(prependedHash, signature);
+
+            if (!_checkSigner(signer)) {
+                return Result.NA;
+            } else if(_isTransfer) {
+                _invalidateSignature(signature);
```

Comment:
instead of `SignatureInvalidated` event, we could emit `SignatureUsed` or something +ve event as transaction get succeed. For  `invalidateSignature()` function this event makes sense but for a successful transfer, it will create confusion.